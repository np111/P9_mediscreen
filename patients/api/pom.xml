<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>com.mediscreen</groupId>
    <artifactId>mediscreen</artifactId>
    <version>1.0-SNAPSHOT</version>
    <relativePath>../../pom.xml</relativePath>
  </parent>
  <artifactId>mediscreen-patients-api</artifactId>
  <packaging>jar</packaging>

  <dependencies>
    <!-- Local modules -->
    <dependency>
      <groupId>com.mediscreen</groupId>
      <artifactId>mediscreen-common-api</artifactId>
    </dependency>

    <!-- Validation API (JSR 380) -->
    <dependency>
      <groupId>javax.validation</groupId>
      <artifactId>validation-api</artifactId>
      <scope>provided</scope>
    </dependency>
  </dependencies>
</project>
