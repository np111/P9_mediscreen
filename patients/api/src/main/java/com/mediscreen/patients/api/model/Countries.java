package com.mediscreen.patients.api.model;

import java.util.Map;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Singular;
import lombok.Value;
import lombok.With;

/**
 * List of countries.
 */
@Builder(builderClassName = "Builder", toBuilder = true)
@Value
@With
@AllArgsConstructor(onConstructor = @__(@Deprecated)) // intended to be used only by tools (MapStruct, Jackson, etc)
public class Countries {
    /**
     * The locale in which the {@linkplain Country#getName() country names} are translated.
     */
    String locale;

    /**
     * The countries, associated with their {@linkplain Country#getCode() code}.
     */
    @Singular("country")
    Map<String, Country> countries;
}
