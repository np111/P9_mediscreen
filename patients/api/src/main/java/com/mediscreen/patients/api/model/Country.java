package com.mediscreen.patients.api.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Value;
import lombok.With;

@Builder(builderClassName = "Builder", toBuilder = true)
@Value
@With
@AllArgsConstructor(onConstructor = @__(@Deprecated)) // intended to be used only by tools (MapStruct, Jackson, etc)
public class Country {
    /**
     * The lower-case ISO 3166-1 alpha-2 country code.
     */
    String code;

    /**
     * The full country name.
     */
    String name;
}
