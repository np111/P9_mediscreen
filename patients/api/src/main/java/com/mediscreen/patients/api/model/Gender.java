package com.mediscreen.patients.api.model;

public enum Gender {
    MALE,
    FEMALE,
}
