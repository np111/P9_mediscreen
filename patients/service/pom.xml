<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>com.mediscreen</groupId>
    <artifactId>mediscreen-common-service-parent</artifactId>
    <version>1.0-SNAPSHOT</version>
    <relativePath>../../common/service-parent/pom.xml</relativePath>
  </parent>
  <artifactId>mediscreen-patients-service</artifactId>
  <packaging>jar</packaging>

  <dependencies>
    <!-- Local modules -->
    <dependency>
      <groupId>com.mediscreen</groupId>
      <artifactId>mediscreen-patients-api</artifactId>
    </dependency>

    <!-- Spring Boot -->
    <!-- * Starter Data JPA (w/ MariaDB) -->
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-data-jpa</artifactId>
    </dependency>
    <dependency>
      <groupId>org.mariadb.jdbc</groupId>
      <artifactId>mariadb-java-client</artifactId>
      <scope>runtime</scope>
    </dependency>

    <!-- Flyway -->
    <dependency>
      <groupId>org.flywaydb</groupId>
      <artifactId>flyway-core</artifactId>
    </dependency>

    <!-- Google libphonenumber -->
    <dependency>
      <groupId>com.googlecode.libphonenumber</groupId>
      <artifactId>libphonenumber</artifactId>
    </dependency>
  </dependencies>
</project>
