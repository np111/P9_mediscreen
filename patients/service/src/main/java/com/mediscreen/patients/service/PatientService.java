package com.mediscreen.patients.service;

import com.mediscreen.patients.api.model.Patient;
import com.mediscreen.patients.exception.PatientNotFoundException;
import java.util.List;
import java.util.UUID;

public interface PatientService {
    List<Patient> list();

    Patient get(UUID id) throws PatientNotFoundException;

    Patient create(Patient patient);

    Patient update(UUID id, Patient patient) throws PatientNotFoundException;

    void delete(UUID id) throws PatientNotFoundException;
}
