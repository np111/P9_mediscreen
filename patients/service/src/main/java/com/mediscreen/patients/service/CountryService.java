package com.mediscreen.patients.service;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mediscreen.patients.api.model.Countries;
import com.mediscreen.patients.api.model.Country;
import com.mediscreen.patients.util.LocalizedMapUtil;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import lombok.Data;
import lombok.Getter;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.stereotype.Service;

@Service
public class CountryService implements InitializingBean {
    private Set<String> countriesCodes;
    private @Getter Set<String> supportedLocales;
    private Countries countriesFallback;
    private Map<String, Countries> countriesByLocale;

    @Override
    public void afterPropertiesSet() throws Exception {
        List<Iso3166Country> iso3166Countries;
        try (Reader rd = new InputStreamReader(CountryService.class.getResourceAsStream("/iso3166.json"), StandardCharsets.UTF_8)) {
            iso3166Countries = new ObjectMapper()
                    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                    .readerForListOf(Iso3166Country.class)
                    .readValue(rd);
        }

        Set<String> countriesCodes = iso3166Countries.stream().map(Iso3166Country::getAlpha2).collect(Collectors.toSet());

        Set<String> supportedLocales = iso3166Countries.stream().flatMap(country -> country.getName().keySet().stream()).collect(Collectors.toSet());

        Map<String, Countries> countriesByLocale = new HashMap<>();
        for (String locale : supportedLocales) {
            Countries.Builder countries = Countries.builder().locale(locale);
            iso3166Countries.forEach(country -> {
                String code = country.getAlpha2();
                String name = country.getName().getOrDefault(locale, country.getName().getOrDefault("en", code));
                countries.country(code, Country.builder().code(code).name(name).build());
            });
            countriesByLocale.put(locale, countries.build());
        }

        this.countriesCodes = Collections.unmodifiableSet(countriesCodes);
        this.supportedLocales = Collections.unmodifiableSet(supportedLocales);
        this.countriesFallback = Objects.requireNonNull(countriesByLocale.get("en"));
        this.countriesByLocale = Collections.unmodifiableMap(countriesByLocale);
    }

    public boolean isCountryCode(String value) {
        return countriesCodes.contains(value);
    }

    public Countries getLocalizedCountries(String locale) {
        return LocalizedMapUtil.get(countriesByLocale, locale, countriesFallback);
    }

    @Data
    private static class Iso3166Country {
        private String alpha2;
        private Map<String, String> name;
    }
}
