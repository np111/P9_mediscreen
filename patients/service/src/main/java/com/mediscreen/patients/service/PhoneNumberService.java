package com.mediscreen.patients.service;

import com.google.i18n.phonenumbers.NumberParseException;
import com.google.i18n.phonenumbers.PhoneNumberUtil;
import com.google.i18n.phonenumbers.Phonenumber.PhoneNumber;
import com.mediscreen.patients.api.model.Country;
import com.mediscreen.patients.api.model.PhoneCountries;
import com.mediscreen.patients.api.model.PhoneCountry;
import com.mediscreen.patients.exception.InvalidPhoneNumberException;
import com.mediscreen.patients.util.LocalizedMapUtil;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Service
public class PhoneNumberService implements InitializingBean {
    private static final Pattern PHONE_NUMBER_PATTERN = Pattern.compile("^[1-9][0-9]{0,3} [1-9][0-9]*$");

    private final PhoneNumberUtil pnu = PhoneNumberUtil.getInstance();
    private final CountryService countryService;

    private PhoneCountries phoneCountriesFallback;
    private Map<String, PhoneCountries> phoneCountriesByLocale;

    @Override
    public void afterPropertiesSet() throws Exception {
        Map<String, PhoneCountries> phoneCountriesByLocale = new HashMap<>();
        for (String locale : countryService.getSupportedLocales()) {
            phoneCountriesByLocale.put(locale, PhoneCountries.builder()
                    .locale(locale)
                    .countries(pnu.getSupportedRegions().stream()
                            .map(region -> PhoneCountry.builder()
                                    .regionCode(region)
                                    .callingCode(pnu.getCountryCodeForRegion(region))
                                    .name(getPhoneCountryName(locale, region))
                                    .build())
                            .sorted(Comparator.comparing(PhoneCountry::getName))
                            .collect(Collectors.toList()))
                    .build());
        }

        this.phoneCountriesFallback = Objects.requireNonNull(phoneCountriesByLocale.get("en"));
        this.phoneCountriesByLocale = Collections.unmodifiableMap(phoneCountriesByLocale);
    }

    private String getPhoneCountryName(String locale, String regionCode) {
        switch (regionCode) {
            case "SH":
                return "Saint Helena";
            case "AC":
                return "Ascension Island";
            case "TA":
                return "Tristan da Cunha";
            case "XK":
                return "Kosovo";
            default:
                Map<String, Country> countries = countryService.getLocalizedCountries(locale).getCountries();
                Country country = countries.get(regionCode.toLowerCase(Locale.ROOT));
                if (country == null) {
                    throw new IllegalArgumentException("Unknown phone country name: " + regionCode);
                }
                return country.getName();
        }
    }

    public PhoneCountries getLocalizedPhoneCountries(String locale) {
        return LocalizedMapUtil.get(phoneCountriesByLocale, locale, phoneCountriesFallback);
    }

    public String parsePhoneNumber(int countryCallingCode, String number) {
        String region = pnu.getRegionCodeForCountryCode(countryCallingCode);
        if ("ZZ".equals(region)) {
            throw new InvalidPhoneNumberException(countryCallingCode, number, "INVALID_COUNTRY_CODE", "Invalid country calling code.");
        }

        PhoneNumber phoneNumber;
        try {
            phoneNumber = pnu.parse(number, region);
        } catch (NumberParseException e) {
            throw new InvalidPhoneNumberException(countryCallingCode, number, e.getErrorType().name(), e.getMessage());
        }

        String ret = phoneNumber.getCountryCode() + " " + phoneNumber.getNationalNumber();
        if (!isValidPhoneNumber(ret)) {
            throw new InvalidPhoneNumberException(countryCallingCode, number, "NOT_A_NUMBER", "The string supplied did not seem to be a phone number.");
        }

        return ret;
    }

    public boolean isValidPhoneNumber(String str) {
        if (!PHONE_NUMBER_PATTERN.matcher(str).matches()) {
            return false;
        }
        int spaceIndex = str.indexOf(' ');
        PhoneNumber phoneNumber = new PhoneNumber();
        phoneNumber.setCountryCode(Integer.parseInt(str.substring(0, spaceIndex)));
        phoneNumber.setNationalNumber(Long.parseLong(str.substring(spaceIndex + 1)));
        return pnu.isValidNumber(phoneNumber);
    }
}
