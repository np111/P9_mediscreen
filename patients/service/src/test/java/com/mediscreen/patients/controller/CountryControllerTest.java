package com.mediscreen.patients.controller;

import com.mediscreen.patients.api.model.Countries;
import com.mediscreen.patients.api.model.Country;
import com.mediscreen.patients.mock.TestConfig;
import com.mediscreen.patients.service.CountryService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.test.web.servlet.MockMvc;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;

@WebMvcTest(CountryController.class)
@Import({TestConfig.class})
class CountryControllerTest extends AbstractControllerTest {
    @MockBean
    private CountryService countryService;

    @Autowired
    private MockMvc mockMvc;

    @Test
    void listCountries() throws Exception {
        doAnswer(m -> Countries.builder()
                .locale(m.getArgument(0))
                .country("us", Country.builder().name("us").name("United States").build())
                .country("fr", Country.builder().name("fr").name("France").build())
                .build())
                .when(countryService).getLocalizedCountries(any());

        toMatchSnapshot(
                mockMvc.perform(get("/country").queryParam("locale", "en")).andReturn(),
                mockMvc.perform(get("/country").queryParam("locale", "fr")).andReturn(),
                mockMvc.perform(get("/country").queryParam("locale", "too_long".repeat(50))).andReturn()
        );
    }
}