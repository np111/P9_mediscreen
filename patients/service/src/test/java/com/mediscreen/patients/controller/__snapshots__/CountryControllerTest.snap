com.mediscreen.patients.controller.CountryControllerTest.listCountries=[
-
> GET /country?locale=en
< HTTP 200
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
"{\"locale\":\"en\",\"countries\":{\"us\":{\"name\":\"United States\"},\"fr\":{\"name\":\"France\"}}}"
-
> GET /country?locale=fr
< HTTP 200
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
"{\"locale\":\"fr\",\"countries\":{\"us\":{\"name\":\"United States\"},\"fr\":{\"name\":\"France\"}}}"
-
> GET /country?locale=too_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_longtoo_long
< HTTP 400
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
"{\"type\":\"CLIENT\",\"code\":\"VALIDATION_FAILED\",\"message\":\"Validation failed: locale size must be between 0 and 255\",\"metadata\":{\"parameter\":\"locale\",\"constraint\":\"Size\",\"min\":0,\"max\":255}}"
]